java/code/src/com/suse/manager/reactor/test/RegisterMinionActionTest.java

```diff
@@ -1905,6 +1911,72 @@ public void testMinionWithUsedReActivationKeyWithStartUpGrains() throws Exceptio
+                    MinionStartupGrains.SuseManagerGrain suseManagerGrain = new MinionStartupGrains.SuseManagerGrain(Optional.of(key));
+                    MinionStartupGrains minionStartUpGrains =  new MinionStartupGrains.MinionStartupGrainsBuilder()
+                            .machineId(MACHINE_ID).saltbootInitrd(false).susemanagerGrain(suseManagerGrain)
+                            .createMinionStartUpGrains();
+                    allowing(saltServiceMock).getGrains(with(any(String.class)), with(any(TypeToken.class)),with(any(String[].class)));
+                    will(returnValue(Optional.of(minionStartUpGrains)));
+                    allowing(saltServiceMock).syncGrains(with(any(MinionList.class)));
+                    allowing(saltServiceMock).syncModules(with(any(MinionList.class)));
+                    allowing(saltServiceMock).getGrains(MINION_ID);
+                    will(returnValue(getGrains(MINION_ID, null, key)));
+                }},
+                (contactMethod) -> {
+                    ActivationKey key = ActivationKeyTest.createTestActivationKey(user);
+                    key.setBaseChannel(baseChannelX8664);
+                    baseChannelX8664.getAccessibleChildrenFor(user)
+                            .forEach(channel -> key.addChannel(channel));
+                    key.setOrg(user.getOrg());
+
+                    Set<ServerGroupType> entitlementsIn = key.getEntitlements();
+                    entitlementsIn.add(ServerConstants.getServerGroupTypeContainerBuildHostEntitled());
+                    key.setEntitlements(entitlementsIn);
+
+                    ActivationKeyFactory.save(key);
+                    return key.getKey();
+                },
+                (optMinion, machineId, key) -> {
+                    assertTrue(optMinion.isPresent());
+                    MinionServer minion = optMinion.get();
+                    assertEquals(MINION_ID, minion.getName());
+
+                    // only channel associated with Activation Key and child
+                    // channels associated with it must be present
+                    assertNotNull(minion.getBaseChannel());
+                    HashSet<Channel> channels = new HashSet<>();
+                    channels.add(baseChannelX8664);
+                    baseChannelX8664.getAccessibleChildrenFor(user)
+                    .forEach(channel -> channels.add(channel));
+                    assertEquals(baseChannelX8664, minion.getBaseChannel());
+                    assertEquals(channels.size(), minion.getChannels().size());
+                    assertTrue(minion.getChannels().containsAll(channels));
+
+                    assertTrue(minion.getFqdns().isEmpty());
+                    // the first ApplyStatesEvent with channels is executed.
+                    // the second (highstate) stay in the queue and can be counted
+                    assertEquals(1, MessageQueue.getMessageCount());
```

Comment:
I would be happy about ideas for better testing it.
I also have no idea why the first state.apply seems to be "executed" and not be anymore in the queue,
but the hightstate stay in the queue.

Reply:
Hm, although it would be nice to have this tested, testing it this way looks quite dangerous to me. The reason, why we only get 1 message here is that the previous `ApplyStatesEvent` was processed by another thread (RHN dispatcher thread). The tests run in "main" thread, which could potentially bring some race conditions (e.g. if you put `Thread.sleep(10000L);` in the `MessageQueue.popEventMessage` on line 144, your test starts failing, as there will be 2 messages in the queue (this simulates the behavior, when the first `ApplyStatesEvent` "would take longer")).

I think that testability of the `MessageQueue` is part of a bigger problem, which should be solved separately and properly, therefore I'd suggest removing this test.

Reply:
removed